### PR TITLE
Fix trailing gap on energy graph cards

### LINKED ISSUE #17389: Trailing space in bar graphs on the energy dashboard
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

The bar graphs on the energy dashboard always have an empty entry at the end.

Examples:
image
image

### Describe the behavior you expected

The last bar within the graph should be closer to the right edge of the card.

### Steps to reproduce the issue

1. Open energy dashboard

### What version of Home Assistant Core has the issue?

core-2023.7.3

### What was the last working version of Home Assistant Core?

_No response_

### In which browser are you experiencing the issue with?

_No response_

### Which operating system are you using to run this browser?

_No response_

### State of relevant entities

_No response_

### Problem-relevant frontend configuration

_No response_

### Javascript errors shown in your browser console/inspector

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by TillFleisch**
## Proposed change

This PR adds a negative offset to the suggested max time of the bar graphs such that the trailing space disappears.
The offset is determined in the same way the tool-tip style is determined and should thus behave correctly.

After:
| After | Before |  
|---|---|
| image  | image |
| image | image |
| image | image |
| image |image |

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #17389
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/lovelace/cards/energy/hui-energy-gas-graph-card.ts [hunk 1]
```diff
@@ -166,7 +166,13 @@ export class HuiEnergyGasGraphCard
...
             type: "time",
             suggestedMin: start.getTime(),
-            suggestedMax: end.getTime(),
+            suggestedMax:
+              end.getTime() -
+              (dayDifference > 35
+                ? 3600000 * 24 * 30
+                : dayDifference > 2
+                ? 3600000 * 24
+                : 3600000),
```
Comment (karwosts) [inline src/panels/lovelace/cards/energy/hui-energy-gas-graph-card.ts:175]: I think you may be removing slightly too much. E.g. for a week display, the start/end goes from Sunday 00:00:00 to Saturday 23:59:59

If you subtract a whole day from the end, then you're now ending on Friday 23:59:59. As a consequence you can see the tick label for Saturday is no longer visible:

image

Maybe use setDate(1), setHour(0), setMinute(0), setSecond(0) ? I think you want the end to fall at the same time as the final datapoint.

Comment (TillFleisch) [reply src/panels/lovelace/cards/energy/hui-energy-gas-graph-card.ts:175]: I agree, using the date methods is also a more elegant solution.

> I think you want the end to fall at the same time as the final datapoint.

Yes. The suggested max (the end) should fall right on the start of the last bar within the chart.

**Comment by karwosts** [comment]
I think this almost looks good but I'm seeing some weird issue that might be related to this in compare data with days of different months. Here I'm comparing April 1-30 against March 1-31. 

Electric grid looks fine, but solar (and gas) does not, the bars are not being spaced correctly. 

image

**Comment by frenck** [comment]
This has already been fixed, thus this PR can be closed.

**Comment by frenck** [comment]
Hmm I discussed this PR with Bram and Paul yesterday. Will check again.

**Pull request closed without merging**